src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Model/EitherTest.cs

```diff
@@ -93,6 +93,165 @@ public void FoldLeft_Right()
+        {
+            var either =
+                new Either<int, string>("either1")
+                    .OnSuccess(() => "either2");
+            
+            Assert.True(either.IsRight);
+            Assert.Equal("either2", either.Right);
+        }
+
+        [Fact]
+        public void OnSuccess_NoArg_Left_PriorFailure()
+        {
+            var eitherResults = new List<string>();
+            
+            var either = 
+                new Either<int, string>(500)
+                    .OnSuccess(() =>
+                    {
+                        eitherResults.Add("either2");
+                        return "either2";
+                    });
+            
+            Assert.True(either.IsLeft);
+            Assert.Equal(500, either.Left);
+            
+            Assert.Empty(eitherResults);
+        }
+
+        [Fact]
+        public async Task OnSuccess_WithEitherTask()
+        {
+            var either =
+                await new Either<int, string>("either1")
+                    .OnSuccess(either1Result => 
+                        Task.FromResult(new Either<int, string>(either1Result + " either2")));
+            
+            Assert.True(either.IsRight);
+            Assert.Equal("either1 either2", either.Right);
+        }
+
+        [Fact]
+        public void OrElse_NoArg()
+        {
+            var either =
+                new Either<int, string>(500)
+                    .OrElse(() => "either2");
+            
+            Assert.True(either.IsRight);
+            Assert.Equal("either2", either.Right);
+        }
+
+        [Fact]
+        public void OrElse_NoArg_Left_PriorSuccess()
+        {
+            var eitherResults = new List<string>();
+            
+            var either = 
+                new Either<int, string>("either1")
+                    .OrElse(() =>
+                    {
+                        eitherResults.Add("either2");
+                        return "either2";
+                    });
+            
+            Assert.True(either.IsRight);
+            Assert.Equal("either1", either.Right);
+            
+            Assert.Empty(eitherResults);
+        }
+
+        [Fact]
+        public async Task OrElse_WithEitherTask()
+        {
+            var either =
+                await new Either<int, string>("either1")
+                    .OnSuccess(either1Result => 
+                        Task.FromResult(new Either<int, string>(either1Result + " either2")));
+            
+            Assert.True(either.IsRight);
+            Assert.Equal("either1 either2", either.Right);
+        }
+
+        [Fact]
+        public async Task OnSuccess_WithEitherTask_Left()
+        {
+            var taskResults = new List<string>();
+            
+            var either =
+                await new Either<int, string>("either1")
+                    .OnSuccess(previousResult => Task.Run(async () =>
+                    {
+                        await Task.Delay(20);
```

Comment:
Yeah I'm adding delays to actually force some async work to be done rather than creating the Tasks that complete immediately, to check that things are being successfully awaited prior to calling the next Task in the chain.  If we always use Task.FromResult or Task.CompetedTask, it's not running anything asynchronously really.  I'm wanting to check that things are executed in the correct order and if they're not, we can easily see which order they're actually being executed in.

Reply:
Note that this is most important within OnSuccess(...) calls that take no inputs and produce more outputs, hence recording their "results" in a List<string>, but I didn't see the harm in using this general approach across the board to make things a little more consistent.

Reply:
Fair enough! Would I be right in thinking that you've also set different delays to make it more obvious if things are being ran in parallel accidentally?

Reply:
Yup indeed!  Longer delays for the earlier Tasks :)